Wrap ilCtrlStructureReader into an Objective

nhaagen opened the pull request:
Use ilCtrlStructureReader with cli-setup's Agent/Objective.
This makes very little changes to the CtrlStructureReader itseltf and thus does not break common setup-routines.

klees commented on Services/UICore/classes/Setup/class.ilCtrlStructureSetupAgent.php:
```diff
@@ -0,0 +1,65 @@
+use ILIAS\Setup;
+use ILIAS\Refinery\Transformation;
+
+class ilCtrlStructureSetupAgent implements Setup\Agent
```
Let's call this `UICoreSetupAgent`. This might want to do something more than just control-structure...

klees commented on Services/Database/classes/Setup/class.ilDatabaseSetupAgent.php:
```diff
@@ -61,7 +61,7 @@ public function getArrayToConfigTransformation() : Transformation {
 		if (!($config instanceof \ilDatabaseSetupConfig)) {
 			throw new \InvalidArgumentException(
-				"Expected \\DatabaseSetupConfig, go '".get_class($config)."' instead."
+				"Expected \\DatabaseSetupConfig, got '".get_class($config)."' instead."
```
:+1:

klees commented on Services/UICore/classes/Setup/class.ilCtrlStructureSetupAgent.php:
```diff
@@ -0,0 +1,65 @@
+	 */
+	public function getUpdateObjective(Setup\Config $config = null): Setup\Objective
+	{
+		return new Setup\NullObjective();
```
```suggestion
		return new \ilCtrlStructureStoredObjective($this->ctrl_reader);
```

I see no harm in reloading the control structure on every update. This will get rid of the requirement that devs need to explicitly do that.

klees commented on Services/UICore/classes/Setup/class.ilCtrlStructureStoredObjective.php:
```diff
@@ -0,0 +1,89 @@
+			throw new \UnachievableException("Need DB to store control-structure");
+		}
+
+		if ($this->isStructureStored($db)) {
```
I think we should just remove this path (and the according method). I rather see this going into the "artifact builder" direction, so this will also not be required in the future (right?).

nhaagen replied:
right.

klees commented on setup/classes/class.ilCtrlStructureReader.php:
```diff
@@ -100,8 +101,11 @@ function readStructure($a_force = false, $a_dir = "", $a_comp_prefix = "",
 	{
-		global $ilDB, $lng;
-		
+		global $lng;
```
Please remove this global as you did with `$ilDB`. Are there any other?

nhaagen replied:
Where am I to pull $lng from? any hints?
btw - lng ist not used but in an exception-message that asks the user to reload the control-structure. Maybe we can reduce to english only in this case and ditch $lng altogether?

This is the line:
383: duplicate_ctrl#:#Error: Duplicate call structure definition found (Class %s) in files:<br /><br />- %s<br />- %s<br /><br />Please remove the file, that does not belong to the official ILIAS distribution.<br />After that invoke 'Tools' -> 'Reload Control Structure' in the ILIAS Setup.###07 09 2006 new variable

alex40724 replied:
Background: It is a common fault of administrators to just unzip ILIAS packages into existing ILIAS code directories. Which leads to the existence of old code files "mostly" overwritten by new code files. But if not included in the new packages, old files will persist. I expect this not only being a problem for the structure reader but also for the artifact reader.

Anyway, I agree that administrators should be able to cope with an English error message, so removing the $lng dependency completely is fine for me.

klees reviewed (approved):
Hi @nhaagen,

I have some minor change requests, but in general this LGTM. I think @alex40724 (as implicit maintainer) should have a look here too.

Best regards!

alex40724 commented:
Thanks!

klees commented:
Thx @alex40724!

nhaagen committed:
Setup: also check on existence of update-file

nhaagen committed:
Setup, DB: fix minor typo

nhaagen committed:
Setup: load Ctrl-Structure via objective
    - read configs for agent-collection into environment in CLI
    - catch potentailly wrong path
    - remove lng from ilSetupLanguage
    - cleanup lang-files
    - use ilUICoreSetupAgent from cli

nhaagen committed:
Setup: extend description of config-file

nhaagen commented:
Hey @klees,
these commits are a bit clearer; is this what you had in mind?

klees commented:
@nhaagen: Yes, thx!
